File: ts/src/day14.ts
import { DayBase } from './day-runner.js';

// https://adventofcode.com/2021/day/14
// See the C# and Python versions for comments

export class Day14 extends DayBase {
    constructor() {
        super(14);
    }

    public async run(): Promise<void> {
        const lines = await this.getInputLines();
        const start = lines[0];
        const inserts = new Map<string, string>();
        for (let i = 2; i < lines.length; i++) {
            let [key, val] = lines[i].split(' -> ');
            inserts.set(key, val);
        }
        console.info(`Day ${this.day}`);
        console.info(`  Part 1:`, this.doInsert(start, inserts, 10));
        console.info(`  Part 2:`, this.doInsert(start, inserts, 40));
    }

    private doInsert(start: string, inserts: Map<string, string>, steps: number): number {
        // split start into pairs
        let pairs = new Map<string, number>();
        for (let i = 0; i < start.length - 1; i++) {
            const k = start[i] + start[i+1];
            pairs.set(k, (pairs.get(k) || 0) + 1);
        }

        for (let step = 0; step < steps; step++) {
            // apply insertion, split into pairs, increment count
            let pairs2 = new Map<string, number>();
            pairs.forEach((v, k) => {
                const toInsert = inserts.get(k) || ' ';
                const k1 = k[0] + toInsert;
                const k2 = toInsert + k[1];
                pairs2.set(k1, (pairs2.get(k1) || 0) + v);
                pairs2.set(k2, (pairs2.get(k2) || 0) + v);
            });
            pairs = pairs2;
        }

        // group and sum
        const counts = new Map<string, number>();
        pairs.forEach((v,k) => {
            counts.set(k[0], (counts.get(k[0]) || 0) + v);
        });
        // don't forget to add the last character of the start string
        counts.set(start[start.length - 1], (counts.get(start[start.length - 1]) || 0) + 1);

        const max = Math.max(...counts.values());
        const min = Math.min(...counts.values());
        return max - min;
    }
}

File: cs/src/day14.cs
public class Day14 : DayBase
{
    public Day14() : base(14) {}

    public override void Run()
    {
        var lines = LoadLines().ToArray();

        var inserts = new Dictionary<string, char>();
        var start = lines[0];
        for (var i = 2; i < lines.Length; i++)
        {
            var splits = lines[i].Split(" -> ");
            inserts[splits[0]] = splits[1][0]; // only single char
        }

        Console.WriteLine($"Day {DayNumber}");
        Console.WriteLine($"  Part 1: {DoInsertions(start, inserts, 10)}");
        Console.WriteLine($"  Part 2: {DoInsertions(start, inserts, 40)}");
    }

    private Int64 DoInsertions(string start, Dictionary<string, char> insertions, int stepCount)
    {
        // Get pairs from start string
        var pairs = new Dictionary<string, Int64>();
        for (var i = 0; i < start.Length - 1; i++)
        {
            var key = start.Substring(i, 2);
            var v1 = pairs.GetValueOrDefault(key);
            pairs[key] = v1 + 1;
        }

        // Each pair of characters will have a single insertion and can then be split
        // into 2 new pairs (ch[0]+insert and insert+ch[1]). Each of those pairs will appear
        // in the new output as many times as the original pair appeared
        for (var step = 1; step <= stepCount; step++)
        {
            var newpairs = new Dictionary<string, Int64>();
            foreach (var (key, value) in pairs)
            {
                var charToInsert = insertions[key];
                var k1 = $"{key[0]}{charToInsert}";
                var k2 = $"{charToInsert}{key[1]}";
                var v1 = newpairs.GetValueOrDefault(k1);
                var v2 = newpairs.GetValueOrDefault(k2);
                newpairs[k1] = v1 + value;
                newpairs[k2] = v2 + value;
            }
            pairs = newpairs;
        }

        // The pairs list will now contain the total number of appearance of each
        // pair of characters. Therefore the total string length will be the sum of
        // the pair counts + 1. The sum of each character will be the sums of each pair
        // where the character appears / 2 - but we also need to add 1 for the start
        // and end characters before halving the sum.
        var counts = new List<(char Letter, Int64 Count)>();
        foreach (var (key, value) in pairs)
        {
            counts.Add((key[0], value));
            counts.Add((key[1], value));
        }
        // Add the start and end characters once each
        counts.Add((start[0], 1));
        counts.Add((start[start.Length - 1], 1));

        // Sum the counts, divide by 2, sort them and take the highest and lowest
        var freq = counts
            .GroupBy(c => c.Letter)
            .Select(g => new { g.Key, Count = g.Sum(c => c.Count) / 2 })
            .OrderBy(x => x.Count);

        var max = freq.Last().Count;
        var min = freq.First().Count;
        return max - min;
    }
}
